[git-webkit] Add Publish Command

JonWBedard opened the pull request:
#### 88545c7a87426658674dbab19b43989e947171e6
<pre>
[git-webkit] Add Publish Command
<a href="https://bugs.webkit.org/show_bug.cgi?id=249586">https://bugs.webkit.org/show_bug.cgi?id=249586</a>
&lt;rdar://97397960&gt;

Reviewed by Elliott Williams.

* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/local/git.py: Add &apos;git branch -a --merged&apos;
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/__init__.py:
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/publish.py: Added.
(Publish):
(Publish.parser): User provides command a list of refs to publish.
(Publish.branches_on): List all branches that are on the history of a specified ref.
(Publish.tags_on): List all tags that are on the history of a specified ref.
(Publish.parental_intersection): Find the intersection point of the provided commit and the parent
branch of that commit. Note that the provided commit might be on the parent branch, in which case
this function returns nothing.
(Publish._push_branch_ref): Push a branch ref onto the provided mapping. This function takes into
consideration any existing branch ref of the same name, taking the more up to date of the two.
(Publish.main): Given the refs provided by the user, find all refs on that history and push those refs
to the &quot;next&quot; source remote.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
(PullRequest.create_pull_request): Prefer the most secret remote.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/publish_unittest.py: Added.
(TestPublish): Added tests.
* metadata/git_config_extension: Sort source remotes.

Canonical link: <a href="https://commits.webkit.org/260044@main">https://commits.webkit.org/260044@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bbf5ca313ab54707d21cb0a3d07618029a53ac3c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106862 "Failed to checkout and rebase branch from PR 7850") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/77/builds/15898 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/39655 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/116043 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/115601 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/11/builds/110765 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/76/builds/17376 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/7096 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/99046 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/19/builds/112629 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/76/builds/17376 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/39655 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/99046 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitpy~~](https://ews-build.webkit.org/#/builders/5/builds/110354 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/76/builds/17376 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/39655 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/99046 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/9049 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/39655 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/9625 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/7096 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [❌ 🧪 services](https://ews-build.webkit.org/#/builders/20/builds/105678 "Failed to checkout and rebase branch from PR 7850") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/15216 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/39655 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/11154 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3747 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/publish.py:
```diff
@@ -0,0 +1,251 @@
+        if output.returncode:
+            sys.stderr.write(output.stderr)
+            return {}
+        result = {None: set()}
```
nit: You might be interested in `collections.defaultdict(set)` to remove some boilerplate.

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/publish.py:
```diff
@@ -0,0 +1,251 @@
+            for commit in commits:
+                prevailing_remote = None
+                remotes_for = repository.branches_for(hash=commit.hash, remote=None).keys()
+                for candidate in repository.source_remotes():
+                    if candidate in remotes_for:
+                        prevailing_remote = candidate
+                        break
+                for candidate in reversed(repository.source_remotes()):
+                    if candidate == args.remote or candidate == prevailing_remote:
+                        args.remote = prevailing_remote
+                        break
```
On my machine, this order appears to be `[origin, apple, fork, origin-fork, apple-fork, emw-apple, emw-apple-apple]`. How is that order guaranteed? How easily can I break things by adding or renaming my remote repos?

Also, why do we have multiple remotes for every repo :sob:

JonWBedard replied:
`source_remotes()` has a `personal` flag which is `false` by default, so that order should be `[origin, apple]` for you, with none of the personal remotes. This order is guaranteed because it's defined in `metadata/git_config_extension`, and the reason I'm changing the order in this PR is so that content will basically go `apple` -> `security` -> `origin`. The only way you could really break assumptions here is if you did something like configuring the remote named `security` to point to something that wasn't https://github.com/WebKit/WebKit-security.

I thought that folks would find having `emw-apple` and `fork` useful (since if I checkout your PRs, my version of your remote would be `emw-apple`), but I think time has proven that thought wrong, will post another PR that discusses unwinding this assumption.

alancoon replied:
I have `security` pointed to something that isn't `WebKit/WebKit-security`

alancoon replied:
What is the usage of the remote being determined in this highlighted logic?
My interpretation is that we're determining which remote to publish to, and this logic is to determine a default, fallback remote in case the user doesn't explicitly provide one, is that correct?

JonWBedard replied:
Your interpretation is correct.

Basically, if the user doesn't specify a remote, we take a look at the fallback list generated here, and determine what the most "secret" remote content is coming from is (most sensitive being `apple`, least sensitive being `origin`). We then publish to the next-most secret remote available to the current user.

JonWBedard replied:
@emw-apple re "why do we have multiple remotes": let's discuss in https://github.com/WebKit/WebKit/pull/8851

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/publish.py:
```diff
@@ -0,0 +1,251 @@
+        return intersection
+
+    @classmethod
+    def _push_branch_ref(cls, mapping, repository, branch, commit):
```
naming nit: This doesn't push anything, could it be called something like `_map_branch_to_push_for_commit`?

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/publish.py:
```diff
@@ -0,0 +1,251 @@
+                encoding='utf-8',
+            ).returncode:
+                sys.stderr.write('Failed to push {} to {}\n'.format(tag, args.remote))
+                return_code += 1
```
Why not push these all in one git invocation, with --atomic, so that the remote isn't left in a indeterminate state if we are interrupted?

JonWBedard replied:
Good idea, always forget that `--atomic` is a thing....

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/publish.py:
```diff
@@ -0,0 +1,251 @@
+        return []
+
+    @classmethod
+    def parental_intersection(cls, repository, commit):
+        if commit.branch == repository.default_branch:
+            return None
+        if commit.identifier <= 1:
+            return None
+        try:
+            parent_branch = repository.find('{}~{}'.format(commit.hash, commit.identifier - 1)).branch
+        except (ValueError, repository.Exception):
+            return None
+        if parent_branch == commit.branch:
+            return None
+        remote = repository.remote_for(parent_branch)
+        intersection = repository.merge_base(commit.hash, 'refs/remotes/{}/{}'.format(remote, parent_branch))
+        if intersection.hash == commit.hash:
+            return None
+        return intersection
```
Could you help me understand why we need to look up _both_ the commit at ID=1, _and_ the merge-base of `commit` + the parent branch? I would think that the commit at ID=1 would _be_ on the parent branch, making the merge-base call unnecessary.

JonWBedard replied:
We have 2 questions we need to answer.

The first is "what is my parent branch". That's actually an extremely difficult question to answer, because `git` doesn't have an idea about which branch came first. People tend to have a concept of this, though, and the WebKit project has some reliable patterns we can use. Generally, we have `main` and then long running branches off of `main`, and then specific release branches off of that long running branch. (sometimes there is actually a 3rd branch between the long running one and the release branches, but we aren't consistent about naming conventions in that case, so it's not easy to determine if two release branches are just sharing a few commits or if we actually have one of these 3rd level branches). Ultimately, we usually care about that long-running branch more than the specific release branch, but our `git-webkit publish` command will almost certainly be handed a branch or a tag that corresponds to a specific release branch. We need to find what the long-running branch associated with the release branch we're on, and `identifier=1` will be the first commit on this long-running branch. If folks are following the right conventions with branch naming, we'll end up getting the long-running branch's name as the "canonical" branch name for that first commit diverging from `main`

The second question we need to answer is "how much of the long-running branch should we be publishing". We can't publish all of the long-running branch, because it's likely that the long-running branch has more commits than just the ones on our release branch, so we only want to publish the long-running branch up to the newest merge-base of the refs the user provided, hence the merge-base. You'll notice that the function returns a commit and not a branch, this is the reason why. We aren't necessarily pushing the entirety of every branch.

emw-apple replied:
Thank you for this!

emw-apple reviewed (commented):
This is too big for me to get a sense of the intricate details. Has it been used to publish a release branch before? Can someone who's done a manual push of release branch content review this?

A little more documentation within the tool and/or the commit message would go a long way. On its face, it's not immediately clear why this is so much more powerful than `git push --follow-tags`

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/publish.py:
```diff
@@ -0,0 +1,255 @@
+        for branch, commit in branches_to_publish.items():
+            print('Pushing {} to {}...'.format(branch, args.remote))
+            if run(
+                [repository.executable(), 'push', args.remote, '{}:refs/heads/{}'.format(commit.hash, branch)],
```
Can we do a single atomic push here too?

Also, curious if you think there's value in showing the git commands we run here. Personally, I would appreciate seeing them for extra confidence that the tool is doing what I think it's doing.

JonWBedard replied:
We can! Thought that we couldn't with branches, for some reason, but it does work.

I will add a logging statement showing the git commands being run, although that logging level isn't enabled by default.

JonWBedard committed:
[git-webkit] Add Publish Command
https://bugs.webkit.org/show_bug.cgi?id=249586
<rdar://97397960>

Reviewed by Elliott Williams.

* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/local/git.py: Add 'git branch -a --merged'
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/__init__.py:
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/publish.py: Added.
(Publish):
(Publish.parser): User provides command a list of refs to publish.
(Publish.branches_on): List all branches that are on the history of a specified ref.
(Publish.tags_on): List all tags that are on the history of a specified ref.
(Publish.parental_intersection): Find the intersection point of the provided commit and the parent
branch of that commit. Note that the provided commit might be on the parent branch, in which case
this function returns nothing.
(Publish._push_branch_ref): Push a branch ref onto the provided mapping. This function takes into
consideration any existing branch ref of the same name, taking the more up to date of the two.
(Publish.main): Given the refs provided by the user, find all refs on that history and push those refs
to the "next" source remote.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
(PullRequest.create_pull_request): Prefer the most secret remote.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/publish_unittest.py: Added.
(TestPublish): Added tests.
* metadata/git_config_extension: Sort source remotes.

Canonical link: https://commits.webkit.org/260044@main
